feat: add --dry-run option

Maxcastel opened the pull request:
Add `--dry-run` option to `doctrine:fixtures:load` command to display loaded fixtures without actually loading them

```bash
php bin/console doctrine:fixtures:load --dry-run
```

closes: #511 #508

Maxcastel committed:
Add --dry-run option

greg0ire commented:
I think this should rather be done by:

1. Contributing a `DryRunORMExecutor` class to `doctrine/data-fixtures`.
2. Using it in this bundle.

It could look like this:

```php
final readonly class DryRunORMExecutor extends AbstractExecutor
{
    use ORMExecutorCommon;

    /** @inheritDoc */
    public function execute(array $fixtures, bool $append = false): void
    {
        $executor = $this;
        $this->em->beginTransaction();
        try {
            if ($append === false) {
                $executor->purge();
            }

            foreach ($fixtures as $fixture) {
                $executor->load($em, $fixture);
            }
            $this->em->flush();
        } finally {
            $this->em->close();
            $this->em->rollBack();
        }
    }
}
```

That way, we avoid nesting transactions (`ORMExecutor` already creates a transaction), and adding to the user interface layer stuff that does not belong to it. Also, the `finally` statement ensures the transaction is cleanly rolled back in the instance of an error.

Maxcastel commented:
Thank you for your response @greg0ire.

So it would be something like this:

`doctrine/doctrine-fixtures-bundle/src/Command/LoadDataFixturesDoctrineCommand.php`:
```php
if ($input->getOption('dry-run')) {
    $ui->text('  <comment>(dry-run)</comment>');
    $executor = new DryRunORMExecutor($em, $purger);
} else {
    $executor = new ORMExecutor($em, $purger);
}
```

If that's the best solution, I'll implement it.

But doesn't that create repetitive code between `DryRunORMExecutor` and `ORMExecutor`?

Wouldn’t it be possible to just modify `ORMExecutor` (for example, change `wrapInTransaction` to a manual transaction and add a boolean `dryRun` parameter) without creating a new `DryRunORMExecutor` class with code similar to `ORMExecutor`?

Something like this:

`doctrine/data-fixtures/src/Executor/ORMExecutor.php`:
```php
final class ORMExecutor extends AbstractExecutor
{
    use ORMExecutorCommon;

    /** @inheritDoc */
    public function execute(array $fixtures, bool $append = false, bool $dryRun = false): void
    {
        $executor = $this;
        $this->em->beginTransaction();
        try {
            if ($append === false) {
                $executor->purge();
            }

            foreach ($fixtures as $fixture) {
                $executor->load($this->em, $fixture);
            }
            $this->em->flush();

            if ($dryRun === true) {
                $this->em->rollBack();
            } else {
                $this->em->commit();
            }
        } catch (\Throwable $e) {   
            $this->em->rollBack();
        } finally {
            $this->em->close();
        }
    }
}
```

`doctrine/doctrine-fixtures-bundle/src/Command/LoadDataFixturesDoctrineCommand.php`:
```php
$executor->execute($fixtures, $input->getOption('append'), $input->getOption('dry-run'));
```

`doctrine/data-fixtures/src/Executor/AbstractExecutor.php`:
```php
/**
 * Executes the given array of data fixtures.
 *
 * @param FixtureInterface[] $fixtures Array of fixtures to execute.
 * @param bool               $append   Whether to append the data fixtures or purge the database before loading.
 * @param bool               $dryRun   Whether to simulate the execution of the fixtures without making actual changes to the database.
 */
abstract public function execute(array $fixtures, bool $append = false, bool $dryRun = false): void;
```

greg0ire commented:
So you would change all the children classes of `AbstractExecutor`?

Maxcastel commented:
> So you would change all the children classes of `AbstractExecutor`?

Yes, or a third parameter `dryRun` defaulting to `false` in the `execute` method of `ORMExecutor`, without changing `execute` in `AbstractExecutor`, to override `execute` in `AbstractExecutor`, or your solution with `DryRunORMExecutor`, what do you prefer?

greg0ire commented:
> But doesn't that create repetitive code between DryRunORMExecutor and ORMExecutor?

Common code could be moved in `ORMExecutorCommon`.

I prefer my solution. If we continue piling up more booleans, the complexity of the method is going to skyrocket.

Maxcastel commented:
@greg0ire
I’ve applied your suggested solution by creating the `DryRunORMExecutor` class in `doctrine/data-fixtures`.
See: [doctrine/data-fixtures PR #552](https://github.com/doctrine/data-fixtures/pull/552)

greg0ire commented:
@Maxcastel can you revive this using the development version of `doctrine/data-fixtures`? That way, once it works, we can release 2.2.0 and merge this.

Maxcastel commented:
@greg0ire 
- rebased on 4.2.x
- switched to doctrine/data-fixtures 2.2.x-dev
- updated the code 
- created new commit to squash

greg0ire commented:
OK, can you try your branch on a real application and confirm that it works?

Maxcastel commented:
- Fixed PHP_CodeSniffer job
- Tested on a real Symfony project with the `dry-run` option and other options, and it works perfectly.
- Added documentation for the `dry-run` option in `docs/index.rst`

greg0ire commented:
Great, I'm going to release 2.2 then.

greg0ire commented:
It will take a bit of time, I'm having issues with deprecations in https://github.com/doctrine/data-fixtures/pull/573

greg0ire commented:
@Maxcastel it's done, you may proceed.

Maxcastel commented:
@greg0ire 
- Rebased on 4.2.x
- Squashed commits
- Tested again on a real Symfony project and it still works

Maxcastel commented:
I just changed the name of commit 19e9ad6

greg0ire commented:
Thanks @Maxcastel !
